src/async.rs

```diff
@@ -5,49 +5,48 @@ use std::{
     time::Duration,
 };
 
-use bytes::Bytes;
-use futures::{Future, IntoFuture};
-use tokio::{net::UdpSocket, util::FutureExt};
+use tokio::net::UdpSocket;
 
-use crate::{handle_response, pdu, SnmpError, SnmpMessageType, SnmpPdu, Value, BUFFER_SIZE};
-
-type SnmpFuture<T> = Box<Future<Item = T, Error = SnmpError> + Send>;
+use crate::{handle_response, pdu, SnmpError, SnmpResult, SnmpMessageType, SnmpPdu, Value, BUFFER_SIZE};
 
 struct AsyncRequest {
     socket: UdpSocket,
-    destination: SocketAddr,
     timeout: Option<Duration>,
 }
 
 impl AsyncRequest {
-    fn send_and_recv(self, pdu: pdu::Buf) -> SnmpFuture<Bytes> {
-        let fut = self
-            .socket
-            .send_dgram(Bytes::from(&pdu[..]), &self.destination)
-            .map_err(|_| SnmpError::SendError);
+    async fn send_and_recv(mut self, pdu: pdu::Buf) -> SnmpResult<Vec<u8>> {
```

Comment:
I am wondering if consuming `self` here is still necessary with async/await.

Reply:
Oh right, thanks :)

Reply:
Would it be useful to change `AsyncSession::req_id` to be AtomicI32 so all methods could take self by shared reference? I have hade some issues with the borrow checker with `.await` in closures with a mutable reference to `AsyncSession`.

Reply:
If `mut` is only needed for `ref_id` then yes, I think it's better to change it.